### ISSUE TITLE
RNN报错

### ISSUE THREAD

**Issue by githwd2016** [zergtant/pytorch-handbook#30]
在第三章的rnn里，如果使用gpu会报错，因为x_np和y_np没有放到gpu上

**Comment by laomao0** [comment]
修改为
rnn.train()
plt.figure(2)
for step in range(EPOCHS):
   start, end = step * np.pi, (step+1)*np.pi # 一个时间周期
   steps = np.linspace(start, end, TIME_STEP, dtype=np.float32)
   x_np = np.sin(steps) 
   y_np = np.cos(steps)
   # GPU
   x_np = torch.tensor(x_np, device=DEVICE).float()
   y_np = torch.tensor(y_np, device=DEVICE).float()
   x = torch.tensor(x_np[np.newaxis, :, np.newaxis]) # shape (batch, time_step, input_size)
   y = torch.tensor(y_np[np.newaxis, :, np.newaxis])
   # cpu
   #x = torch.from_numpy(x_np[np.newaxis, :, np.newaxis]) # shape (batch, time_step, input_size)
   #y = torch.from_numpy(y_np[np.newaxis, :, np.newaxis])
   prediction, h_state = rnn(x, h_state) # rnn output
   # 这一步非常重要
   h_state = h_state.data # 重置隐藏层的状态, 切断和前一次迭代的链接
   loss = criterion(prediction, y) 
   # 这三行写在一起就可以
   optimizer.zero_grad() 
   loss.backward() 
   optimizer.step() 
   if (step+1)%20==0: #每训练20个批次可视化一下效果，并打印一下loss
       print("EPOCHS: {},Loss:{:4f}".format(step,loss))
       # copy to cpu
       y_np_np = y_np.cpu().numpy()
       prediction_np = prediction.cpu()
       plt.plot(steps, y_np_np.flatten(), 'r-')
       plt.plot(steps, prediction_np.data.numpy().flatten(), 'b-')
       plt.draw()
       plt.pause(0.01)

**Comment by laomao0** [comment]
`rnn.train()
plt.figure(2)
for step in range(EPOCHS):
   start, end = step * np.pi, (step+1)*np.pi # 一个时间周期
   steps = np.linspace(start, end, TIME_STEP, dtype=np.float32)
   x_np = np.sin(steps) 
   y_np = np.cos(steps)
   # GPU
   x_np = torch.tensor(x_np, device=DEVICE).float()
   y_np = torch.tensor(y_np, device=DEVICE).float()
   x = torch.tensor(x_np[np.newaxis, :, np.newaxis]) # shape (batch, time_step, input_size)
   y = torch.tensor(y_np[np.newaxis, :, np.newaxis])
   # cpu
   #x = torch.from_numpy(x_np[np.newaxis, :, np.newaxis]) # shape (batch, time_step, input_size)
   #y = torch.from_numpy(y_np[np.newaxis, :, np.newaxis])
   prediction, h_state = rnn(x, h_state) # rnn output
   # 这一步非常重要
   h_state = h_state.data # 重置隐藏层的状态, 切断和前一次迭代的链接
   loss = criterion(prediction, y) 
   # 这三行写在一起就可以
   optimizer.zero_grad() 
   loss.backward() 
   optimizer.step() 
   if (step+1)%20==0: #每训练20个批次可视化一下效果，并打印一下loss
       print("EPOCHS: {},Loss:{:4f}".format(step,loss))
       #copy to cpu
       y_np_np = y_np.cpu().numpy()
       prediction_np = prediction.cpu()
       plt.plot(steps, y_np_np.flatten(), 'r-')
       plt.plot(steps, prediction_np.data.numpy().flatten(), 'b-')
       plt.draw()
       plt.pause(0.01)`
